### PR TITLE
Staging the vSphere Cloud Provider

### PR DESCRIPTION
**Pull request by andrewsykim**
Signed-off-by: Andrew Sy Kim <kiman@vmware.com>

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Staging the vSphere Cloud Provider as part of KEP [20190125-removing-in-tree-providers](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/20190125-removing-in-tree-providers.md). Staging repo setup here https://github.com/kubernetes/legacy-cloud-providers

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: hack/verify-publishing-bot.py [hunk 1]
```diff
@@ -48,6 +48,9 @@ def main():
     rootdir = os.path.dirname(__file__) + "/../"
     rootdir = os.path.abspath(rootdir)
 
+    # staging directories that do not need to be published yet
+    publishing_exceptions = ["cloud-provider-vsphere"]
```
Comment (andrewsykim) [inline hack/verify-publishing-bot.py]: cc @nikhita 

Adding the vsphere staging directory as an exception for now until we sort out https://github.com/kubernetes/publishing-bot/issues/156

**Comment by frapposelli** [comment]
Thanks @andrewsykim for initiating this 👍 

I'm absolutely in favor of the change, can you clarify the impact of kubernetes/publishing-bot#156 once it's merged? I'm particularly interested on the effects on existing code living at k8s.io/cloud-provider-vsphere

Also, this will require in-depth manual testing before release, so
/hold
for now.

**Comment by andrewsykim** [comment]
The intention there is to publish a subdirectory of a staged repo (`staging/src/k8s.io/cloud-provider-vsphere`) to a subdirectory of the corresponding external repo (`k8s.io/cloud-provider-vsphere`). For this case, that would mean only publishing `staging/src/k8s.io/cloud-provider-vsphere/legacyprovider` -> `k8s.io/cloud-provider-vsphere/legacyprovider` where only the `legacyprovider` is published to the vSphere external repo but everything else is updated directly from it's own tree. In other words, do not touch the `legacyprovider` package outside of this staging directory.

The main justification for this is allowing providers to opt into the in-tree implementation for their out-of-tree provider without having to maintain 2 versions of the provider in different trees. This is completely optional though in which case the staged legacy provider will just be removed in a future release.

The implementation details around how this will be done in the publishing bot is still being worked on. I expect some details to change as we move from Godeps to Go modules and some changes required in the publishing bot to support subdirectory publishing (vs entire repos). Having said that, I don't think the publishing bot details/implementation should block this PR since the legacy provider is imported and functions exactly as it was before, the only change is where the package lives. Nothing will be published to the external repo until we update the [publishing rules](https://github.com/kubernetes/kubernetes/blob/master/staging/publishing/rules.yaml) which I intend to do in a follow-up PR.

**Comment by frapposelli** [comment]
@andrewsykim that was exhaustive, thanks a lot! :+1:

/lgtm
/approve

File: hack/verify-publishing-bot.py (hunk 1 above)
Comment (nikhita) [reply hack/verify-publishing-bot.py]: /cc @dims @sttts

**Comment by andrewsykim** [comment]
Thanks Fabio! Given go modules support is going to merge soon, I'm going to hold on this PR for a bit

**Comment by sttts** [comment]
How do you think https://github.com/kubernetes/publishing-bot/issues/156 will be addressed?

File: hack/verify-publishing-bot.py (hunk 1 above)
Comment (sttts) [reply hack/verify-publishing-bot.py]: What does it imply to be on this list? This must be documented.

**Comment by andrewsykim** [comment]
TBH I don't know yet :). I'd like to see what the upcoming go modules changes look like to get a better idea of how to address this. Will follow-up in a few days

File: hack/verify-publishing-bot.py (hunk 1 above)
Comment (andrewsykim) [reply hack/verify-publishing-bot.py]: I don't think we'll need this anymore, proposed changes in https://github.com/kubernetes/enhancements/pull/952

**Comment by nikhita** [comment]
Two nits:

- Can you add this repo to the README here: https://github.com/kubernetes/kubernetes/tree/master/staging?
- The main PR body says `Note that we are not publishing the staging directory into any external repos until kubernetes/publishing-bot#156 is addressed.` This is not true anymore, right? Can you update it? Just confused me for a second. :)

**Comment by andrewsykim** [comment]
/assign @thockin
/hold

Holding until publishing bot is setup but the PR is ready for approval/review.

**Comment by andrewsykim** [comment]
/hold cancel

Looks like publishing bot is setup, https://github.com/kubernetes/publishing-bot/pull/176 can merge after this PR.

**Review by thockin — COMMENTED** [review]
@liggitt do we still need the symlinks in vendor/... ?

/lgtm
/approve

**Comment by liggitt** [comment]
This needs a rebase on master to pick up a few dep bumps that happened in the last week. I expect verify-dependencies would fail on it in its current state, since the new repo has a stale version of ginkgo pinned relative to master. Let's find out. 

/test pull-kubernetes-dependencies

> @liggitt do we still need the symlinks in vendor/... ?

Yes, we're still using vendor for now, and those are maintained by `hack/update-vendor.sh`

**Comment by liggitt** [comment]
Yup

```diff
-	k8s.io/klog v0.0.0-20190306015804-8e90cee79f82
+	k8s.io/klog v0.3.0
 )
 
 replace (
-	github.com/onsi/ginkgo => github.com/onsi/ginkgo v0.0.0-20170318221715-67b9df7f55fe
…
```

**Comment by andrewsykim** [comment]
@liggitt rebased and updated, can I get a lgtm please :)

**Comment by thockin** [comment]
I guess I assumed that a `replace` directive would apply before a lookup in
vendor/.  E.g. `go build` would resolve "k8s.io/legacy-cloud-providers" to
"./staging/src/k8s.io/legacy-cloud-providers" and therefore not need the
vendor symlinks.  Is that not true?

**Comment by thockin** [comment]
Uggh, it looks like code generators fail because they assume vendor and are
not module-smart.  Objection waived for now.  Jordan, we should queue that
up in an issue or something.

**Pull request merged**